src/mscorlib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -628,6 +628,49 @@ public bool Remove(TKey key)
             return false;
         }
 
+        public bool Remove(TKey key, out TValue value)
+        {
+            if (key == null)
+            {
+                value = default(TValue);
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.key);
+            }
+
+            if (buckets != null)
```

Comment:
@safern I believe this is in the pursuit of reducing duplicated code. But then the duplication, although restricted to few lines, happens between `FindEntry` and `FindEntryAndRemove` for the _finding_ part. Unless we want to merge them and pass a `boolean` to indicate 'remove' that will default to `false`...? ... potentially introducing a branch on the paths where existing `FindEntry` is used. Either ways, the method `FindEntryAndRemove` still _feels_ incomplete from design point of view because we have removed but not free'd the entry and freeing part comes as post fix...

@jkotas 
1. Curiosity question, wouldn't the C# compiler perform local CSE on the four lines which fetch the entry ... i.e. `entry[i].*` lines... with effect being that the entry is computed only once? This is based on my experience with C / C++ compilers though...
1. How does the usage of `default(TValue)` work? Does it initialize and then do a memberwise copy to destination (due to operator =) or, does it directly construct the default into memory referenced by the destination? If it is latter, does it still do memberwise init or it does sort of memset to zeros using span APIs? Thought in my head: cycles spent in memberwise init to default for a value type will be proportional to its size.